## Installation ##
Command: git clone https://github.com/lfdelzam/primer_desing.git

The pipeline uses the programes:

[snakemake](https://snakemake.github.io) 3.13.3

[muscle](http://www.drive5.com/muscle) v3.8.1551

[Degeprime](https://github.com/andand/DEGEPRIME)

[mfeprimer](https://github.com/quwubin/MFEprimer-3.0/releases/tag/v3.2.2) 3.2.2
	
## Create conda environment ##

conda create -n primers -c bioconda blast=2.5.0 snakemake=3.13.3 krona=2.7.1 muscle=3.8.1551

## Usage ##

set pipeline parameters in config_primer_design.json using the command: `nano config_primer_design.json`, 
modify the parameters and save changes by taping `ctrl x` and tape `y`

  "work_dir": "/absolute/path/to/primer_design/",
  
  "tar_gff_files": "/absolute/pat/to/genome_gff_files.tar",   -- tar file downloaded from NCBI --
  
  "tar_file_genomes": "/absolute/path/to/genome_fasta.tar",   -- tar file downloaded from NCBI --
  
  "NCBI_gene_name": "your option", -- NCBI name of the target gene (required) --
  
  "gff_file": "/absolute/path/to/gff_with_selected_gene_annotation.gff", -- if not provided, it will be generated using "tar_gff_files", "tar_file_genomes", "NCBI_gene_name" --
  
  "file_suffix_name": "your option",
  
  "gene_sequences": "/absolute/path/to/gene_sequeneces.fna",  -- if not provided it will be generated using "tar_gff_files", "tar_file_genomes", and "gff_file" --
  
  "muscle_params": "-maxiters 3",
  
  "trim_degeprime": "-min 0.9",
  
  "degenerecies": "4,8,12,18,20,24,32,48,60,80,92,120",
  
  "primer_sizes": "17,18,19,20,21,22,23,24,25,26",
  
  "primer_selection_parameters": "-c 0.995 -g 35", -- -c stands for coverage (fraction) and -g for minimum GC content (%) --
  
   "amplicon_size": "200,1000" -- minimum and maximum amplicon size --


command: 
`conda activate primers`

run the pipeline using the commnad: 
`snakemake -s design_primer --cores <number of threads>`

## output ##

Directory `Results`:

	file "List_<file_suffix_name>": Lists all possible primer paire combinations of selected primers	
		
	file "primers_<file_suffix_name>_table": Lists of selected primers. It includes primer name, sequences, reverse complement, degenerecy, length, GC range, TM range


# intermediare output #

Directory `MSA`: 

It contains the multiple sequence alignment generated by muscle file `MSA_<file_suffix_name>`, and the file `timmed_align_file_from_MSA_<file_suffix_name>` generated by Degeprime

Directory `Potential_primers`: 

It contains the files generated by MFEprimer (dimer and hairpin primer evaluation)

